ruma-common/src/push/condition.rs

```diff
@@ -42,13 +49,203 @@ pub enum PushCondition {
+                    Some(v) => match UserId::try_from(v.as_str()) {
+                        Ok(u) => u,
+                        Err(_) => return false,
+                    },
+                    None => return false,
+                };
+
+                let sender_level = match context.users_power_levels.get(&sender_id) {
+                    Some(i) => i,
+                    None => &context.default_power_level,
+                };
+
+                sender_level >= context.notification_power_levels.get(key)
+            }
+        }
+    }
+}
+
+/// The context of the room associated to an event to be able to test all push conditions
+#[derive(Clone, Debug)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct PushConditionRoomCtx {
+    /// The roomId of the room
+    pub room_id: String,
+
+    /// The number of members in the room
+    pub member_count: UInt,
+
+    /// The display name of the current user in the room
+    pub user_display_name: String,
+
+    /// The power levels of the users of the room
+    pub users_power_levels: BTreeMap<UserId, Int>,
+
+    /// The default power level of the users of the room
+    pub default_power_level: Int,
+
+    /// The notification power levels of the room
+    pub notification_power_levels: NotificationPowerLevels,
+}
+
+/// Matches the given value against a pattern.
+///
+/// The match is case insensitive.
+///
+/// If `match_words` is true, looks for pattern as a substring of `value`,
+/// and checks that it is a "word". If it is false, checks the pattern as
+/// a glob with wildcards `*` and `?`.
+fn matches_pattern(value: &str, pattern: &str, match_words: bool) -> bool {
+    if value.is_empty() || pattern.is_empty() {
+        return false;
+    }
+
+    let value = &value.to_uppercase();
+    let pattern = &pattern.to_uppercase();
+
+    if match_words {
+        matches_word(value, pattern)
+    } else {
+        WildMatch::new(pattern).matches(value)
+    }
+}
+
+fn matches_word(value: &str, pattern: &str) -> bool {
+    match value.find(pattern) {
+        Some(start) => {
+            let end = start + pattern.len();
+
+            // Look if the match has word boundaries
+            let word_boundary_start = {
+                if !char_at(value, start).is_alphanumeric() {
+                    true
+                } else {
+                    match find_prev_char(value, start) {
+                        Some(c) => !c.is_alphanumeric(),
+                        None => true,
+                    }
+                }
+            };
+
+            if word_boundary_start {
+                let word_boundary_end = {
+                    if end == value.len() || !find_prev_char(value, end).unwrap().is_alphanumeric()
+                    {
+                        true
+                    } else {
+                        let c = char_at(value, end);
+                        !c.is_alphanumeric()
+                    }
+                };
```

Comment:
The whole `if cond { true } else { let c = char_at(value, end); !c.is_alphanumeric() }` can be shortened to `cond || !char_at(value, end).is_alphanumeric()`, right?